name: UGC_service_2

on:
  push:
    branches:
      - main
  pull_request:
    branches:
      - main

jobs:
  build:

    runs-on: ubuntu-latest

    continue-on-error: true
    strategy:
      matrix:
        python-version: [3.7, 3.8, 3.9]

    steps:
      - uses: actions/checkout@v3

      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@v3
        with:
          python-version: ${{ matrix.python-version }}

      - name: Install dependencies on Python ${{ matrix.python-version }}
        run: |
          python -m pip install --upgrade pip
          pip install wemake-python-styleguide flake8-html mypy lxml
          if [ -f requirements.txt ] && ["${{ matrix.python-version }}" != "3.7" ];
            then pip install -r requirements.txt;
          fi

      - name: Run Flake8
        id: run_flake8
        run: flake8
              --config=setup.cfg
              --exclude=research,elk/main.py
              --format=html
              --htmldir=flake-report .
        continue-on-error: true

      - name: Save flake8 HTML report
        uses: actions/upload-artifact@v2
        continue-on-error: true
        with:
          name: flake-report
          path: flake-report/

      - name: Run MyPy
        id: run_mypy
        run: |
            mkdir mypy-report
            mypy --config-file ./setup.cfg \
            --exclude "^elk/main\.py$" \
            --exclude "^research/" \
            --show-error-codes \
            --html-report mypy-report .
        continue-on-error: true

      - name: Save mypy HTML report
        uses: actions/upload-artifact@v2
        continue-on-error: true
        with:
          name: mypy-report
          path: mypy-report/

      - name: Fail if errors found
        run: |
          if [ "${{ steps.run_flake8.outcome }}" == "failure" ]; then
            echo "Linting errors found. Exiting with error."
            exit 1
          fi

          if [ "${{ steps.run_mypy.outcome }}" == "failure" ]; then
            echo "MyPy errors found. Exiting with error."
            exit 1
          fi

      - name: Telegram Notification
        if: cancelled() == false
        uses: appleboy/telegram-action@master
        continue-on-error: true
        with:
          to: ${{ secrets.TELEGRAM_CHAT_ID }}
          token: ${{ secrets.TELEGRAM_BOT_TOKEN }}
          message: |
            Repository: ${{ github.repository }}
            Event: ${{ github.event_name }}
            From user: ${{ github.actor }}
            Branch: ${{ github.ref }}
            Python version: ${{ matrix.python-version }}
            status: ${{ job.status }}
